openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,102 @@
+Discussion provider configuration
+=================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+As part of the BD-03 initiative (Blended Development, Project 3),
+we want to enable third-party discussion providers to replace the
+default forums experience in edx-platform.
+
+To accomplish this, we should establish a configuration system to
+enable/disable/configure these new discussion-provider plugins;
+the existing/default/legacy forums experience will be included as one of these plugins.
+
+This ADR proposes a new configuration system that allows operators to create
+pre-populated configurations for specific discussions providers,
+which can then be used and customized by course authors/admins.
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model with
+historical records, `LearningContextDiscussionConfig`.
+
+While the underlying model will use `LearningContextKeys` (so that this
+could eventually be extended to cover (sites/programs/etc.), the first
+implementing data structure (`data.py`) can be described as:
+
+    @attr.s(frozen=True)
+    class CourseDiscussionConfigData:
+        course_key = attr.ib(type=CourseKey)
+        enabled = attr.ib(type=bool)
+        lti_configuration = attr.ib(type=dict)
+        plugin_configuration = attr.ib(type=dict)
+        provider = attr.ib(type=str)
```

Comment:
Thinking out loud - This provider ID uniquely identifies an entry in our hard-coded (for the foreseeable future) list of of discussions apps.  It could be used to show the provider's name (Piazza, edX Forums) and discover what capabilities it has.  These IDs must be unique across providers.

Reply:
Long term, if we expand this pattern of app providers, we could move the provider definition to a DB table and this becomes a foreign key.

Reply:
Yeah, in the ADR I think I referred to it as "arbitrary, but stable". :+1: on future considerations for storage standardization.